### PR TITLE
Issuer Sign Tool extention support

### PR DESCRIPTION
**Pull request by NMorozxov**
Issuer Sign Tool (1.2.643.100.112) The name of the tool used to signs the subject (ASN1_SEQUENCE)
This extention is required to obtain the status of a qualified certificate at Russian Federation.
RFC-style description is available here: https://tools.ietf.org/html/draft-deremin-rfc4491-bis-04#section-5
Russian Federal Law 63 "Digital Sign" is available here:  http://www.consultant.ru/document/cons_doc_LAW_112701/

##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by lumag** [comment]
An example certificate with these extensions

[grfc.txt](https://github.com/openssl/openssl/files/4274755/grfc.txt)

**Comment by beldmit** [comment]
@NMorozxov, Could you also implement a test using the provided certificate?

**Comment by NMorozxov** [comment]
Is there any example for the test?

**Comment by lumag** [comment]
> Is there any example for the test?

https://github.com/openssl/openssl/files/4274755/grfc.txt

**Comment by NMorozxov** [comment]
Dmitry is prev PR says that he can help me to convert a cert into a test. i want an example of converting, you cert exapmle i have seen.

**Comment by beldmit** [comment]
You can make smth similar to 
https://github.com/openssl/openssl/blob/master/test/recipes/25-test_x509.t

**Comment by NMorozxov** [comment]
ping @beldmit

**Comment by beldmit** [comment]
So are you able to extend some existing x509 tests with the provided example? If not, I can assist

**Comment by NMorozxov** [comment]
> So are you able to extend some existing x509 tests with the provided example? If not, I can assist

I have added a new test. You want remove it and add to existig one? Ok i can do it.

**Comment by beldmit** [comment]
My fault. Missed the extra commit.

Could you please resolve the conflict (hint: it's better to move newly added functions in `util/libcrypto.num` 3 lines up to avoid them in future).

**Comment by NMorozxov** [comment]
> My fault. Missed the extra commit.
> 
> Could you please resolve the conflict (hint: it's better to move newly added functions in `util/libcrypto.num` 3 lines up to avoid them in future).

done

**Comment by beldmit** [comment]
travis failure is a timeout

**Comment by NMorozxov** [comment]
> travis failure is a timeout

sometimes happens....

**Comment by NMorozxov** [comment]
ping

**Comment by NMorozxov** [comment]
Ping @openssl/otc

File: crypto/x509/v3_ist.c [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+IMPLEMENT_ASN1_FUNCTIONS(ISSUER_SIGN_TOOL)
+
+
+static ISSUER_SIGN_TOOL *v2i_issuer_sign_tool(X509V3_EXT_METHOD *method, X509V3_CTX *ctx,
+                        STACK_OF(CONF_VALUE) *nval)
+{
+    ISSUER_SIGN_TOOL *ist = ISSUER_SIGN_TOOL_new();
+    int i;
+    if (ist == NULL) {
```
Comment (t8m) [inline crypto/x509/v3_ist.c]: Empty line missing after declaration.

File: crypto/x509/v3_ist.c [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        X509V3err(X509V3_F_V2I_ISSUER_SIGN_TOOL, ERR_R_MALLOC_FAILURE);
+        return NULL;
+    }
+    for (i = 0; i < sk_CONF_VALUE_num(nval); ++i) {
+        CONF_VALUE *cnf = sk_CONF_VALUE_value(nval, i);
+        if (strcmp(cnf->name, "signTool")) {
```
Comment (t8m) [inline crypto/x509/v3_ist.c]: Empty line after declaration missing. Also isn't the condition for strcmp inverted? Shouldn't it be strcmp() == 0?

File: test/recipes/25-test_rusext_data/grfc.msb [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        Validity
+            Not Before: Mar 12 07:38:26 2013 GMT
+            Not After : Mar 12 07:46:00 2028 GMT
+        Subject: OGRN=1027739334479, INN=007706228218, street=\U0414\U0435\U0440\U0431\U0435\U043D\U0435\U0432\U0441\U043A\U0430\U044F \U043D\U0430\U0431. \U0434. 7 \U0441\U0442\U0440. 15, emailAddress=pki-grfc@grfc.ru, C=RU, ST=77 \U0433. \U041C\U043E\U0441\U043A\U0432\U0430, L=\U041C\U043E\U0441\U043A\U0432\U0430, O=\U0424\U0413\U0423\U041F "\U0413\U0420\U0427\U0426", CN=\U0423\U0426 \U0424\U0413\U0423\U041F "\U0413\U0420\U0427\U0426"
+        X509v3 extensions:
+            Signing Tool of Subject: 
+                "КриптоПро CSP" (версия 3.6)
+            Signing Tool of Issuer: 
+                signTool:     "КриптоПро CSP" (версия 3.6)
```
Comment (t8m) [inline test/recipes/25-test_rusext_data/grfc.msb]: As you can see above the padding in Validity fields is done before the :. Please consider following this.

File: crypto/x509/v3_ist.c [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+        }
+    }
+    return ist;
+}
+
+static int i2r_issuer_sign_tool(X509V3_EXT_METHOD *method,
+                                 ISSUER_SIGN_TOOL *ist, BIO *out,
+                                 int indent)
+{
+    if (ist->signTool) {
```
Comment (t8m) [inline crypto/x509/v3_ist.c]: The tests for nonNULL must be explicit - ist->signTool != NULL and similarly below.

File: crypto/x509/v3_ist.c [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+        CONF_VALUE *cnf = sk_CONF_VALUE_value(nval, i);
+        if (strcmp(cnf->name, "signTool")) {
+            ist->signTool = ASN1_UTF8STRING_new();
+            if (ist->signTool == NULL) {
+                X509V3err(X509V3_F_V2I_ISSUER_SIGN_TOOL, ERR_R_MALLOC_FAILURE);
+                ISSUER_SIGN_TOOL_free(ist);
+                return NULL;
+            }
+            ASN1_STRING_set(ist->signTool, cnf->value, strlen(cnf->value));
+        } else if (strcmp(cnf->name, "cATool")) {
```
Comment (t8m) [inline crypto/x509/v3_ist.c]: And here and below...

File: crypto/x509/v3_ist.c (hunk 1 above)
Comment (NMorozxov) [reply crypto/x509/v3_ist.c]: fixed

File: crypto/x509/v3_ist.c (hunk 2 above)
Comment (NMorozxov) [reply crypto/x509/v3_ist.c]: fixed

File: test/recipes/25-test_rusext_data/grfc.msb (hunk 3 above)
Comment (NMorozxov) [reply test/recipes/25-test_rusext_data/grfc.msb]: ok, i use indent but the result change is minor

File: crypto/x509/v3_ist.c (hunk 4 above)
Comment (NMorozxov) [reply crypto/x509/v3_ist.c]: fixed

File: crypto/x509/v3_ist.c (hunk 5 above)
Comment (NMorozxov) [reply crypto/x509/v3_ist.c]: fixed

File: test/recipes/25-test_rusext_data/grfc.msb (hunk 3 above)
Comment (NMorozxov) [reply test/recipes/25-test_rusext_data/grfc.msb]: ohh, i understand ehat you want :) 
fixed

**Comment by NMorozxov** [comment]
ping @beldmit

**Comment by NMorozxov** [comment]
who will submit this?

**Comment by beldmit** [comment]
If everything is OK, I will

**Comment by NMorozxov** [comment]
What blocks submitting?

**Comment by beldmit** [comment]
First of all, we have 24 hours of commit delay from the moment the PR was approved.

Second, I must have some free time to commit.

**Comment by NMorozxov** [comment]
ok :)

**Comment by beldmit** [comment]
Merged. Many thanks!

**Pull request closed without merging**